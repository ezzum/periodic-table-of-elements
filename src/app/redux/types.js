export const ELEMENT_MOUSE_ENTER = 'ELEMENT_MOUSE_ENTER';
export const ELEMENTS_LOAD = 'ELEMENTS_LOAD';
export const ELEMENT_CLICK = 'ELEMENT_CLICK';
export const FILTER_BUTTON_CLICK = 'FILTER_BUTTON_CLICK';
export const SHOW_SPINNER = 'SHOW_SPINNER';
export const HIDE_SPINNER = 'HIDE_SPINNER';
