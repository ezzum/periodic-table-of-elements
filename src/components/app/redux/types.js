export const ELEMENT_MOUSE_ENTER = 'ELEMENT_MOUSE_ENTER';
export const ELEMENTS_LOAD = 'ELEMENTS_LOAD';
export const ELEMENT_CLICK = 'ELEMENT_CLICK';
export const FILTER_BUTTON_CLICK = 'FILTER_BUTTON_CLICK';