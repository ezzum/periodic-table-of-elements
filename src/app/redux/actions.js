import axios from 'axios';
import {
  ELEMENTS_LOAD,
  ELEMENT_CLICK,
  ELEMENT_MOUSE_ENTER,
  FILTER_BUTTON_CLICK,
  HIDE_SPINNER,
  SHOW_SPINNER,
  SEARCH_ELEMENT,
} from './types';

export function showSpinner() {
  return {
    type: SHOW_SPINNER,
  };
}

export function hideSpinner() {
  return {
    type: HIDE_SPINNER,
  };
}

export function ElementsLoad() {
  return async (dispatch) => {
    try {
      dispatch(showSpinner());
      const url = 'https://data-base-chem.herokuapp.com/table';
      const response = await axios.get(url);
      const elements = response.data.elements.map((item) => {
        return { ...item, active: false };
      });
      dispatch({ type: ELEMENTS_LOAD, value: elements });
      dispatch(hideSpinner());
    } catch (error) {
      return null;
    }
  };
}

export function ElementMouseEnter(value) {
  return {
    type: ELEMENT_MOUSE_ENTER,
    value,
  };
}

export function ElementClick(value) {
  return {
    type: ELEMENT_CLICK,
    value,
  };
}

export function FilterButtonClick(value) {
  return {
    type: FILTER_BUTTON_CLICK,
    value,
  };
}

export function SearchElement(value) {
  return {
    type: SEARCH_ELEMENT,
    value,
  };
}
