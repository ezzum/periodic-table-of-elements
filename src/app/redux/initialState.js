export const initialState = {
  elements: [],
  activeElement: '',
  previewElement: {
    name: 'Наведите курсор',
    group: 'на любой',
    symbol: 'элемент',
  },
  tableProps: {
    columns: 18,
    rows: 7,
  },
  activeButton: {
    metal: false,
    semiMetal: false,
    nonMetal: false,
  },
  loading: false,
};
