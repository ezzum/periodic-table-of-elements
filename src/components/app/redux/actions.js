import { ELEMENTS_LOAD, ELEMENT_CLICK, ELEMENT_MOUSE_ENTER, FILTER_BUTTON_CLICK } from "./types";

export function ElementsLoad(value) {
    return {
        type: ELEMENTS_LOAD,
        value
    }
}

export function ElementMouseEnter(value) {
    return {
        type: ELEMENT_MOUSE_ENTER,
        value
    }
}

export function ElementClick(value) {
    return {
        type: ELEMENT_CLICK,
        value
    }
}

export function FilterButtonClick(value) {
    return {
        type: FILTER_BUTTON_CLICK,
        value
    }
}