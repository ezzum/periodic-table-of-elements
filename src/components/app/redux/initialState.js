export const initialState = {
    elements: [],
    activeElement: '',
    previewElement: '',
    tableProps: {
        columns: 18,
        rows: 7
    },
    activeButton: {
        metal: false,
        semiMetal: false,
        nonMetal: false
    }
};