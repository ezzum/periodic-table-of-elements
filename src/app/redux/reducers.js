import {
  ELEMENTS_LOAD,
  ELEMENT_CLICK,
  ELEMENT_MOUSE_ENTER,
  FILTER_BUTTON_CLICK,
  HIDE_SPINNER,
  SHOW_SPINNER,
  SEARCH_ELEMENT,
  GET_PAGE_ID,
  GET_PAGE_CONTENT,
  CLEAR_DESCRIPTION,
} from './types';
import { initialState } from './initialState';

export function rootReducer(state = initialState, action) {
  switch (action.type) {
    case ELEMENTS_LOAD:
      return {
        ...state,
        elements: action.value,
      };
    case ELEMENT_MOUSE_ENTER:
      return {
        ...state,
        previewElement: action.value,
      };
    case ELEMENT_CLICK:
      return {
        ...state,
        activeElement: {
          ...state.activeElement,
          name: action.value,
        },
      };
    case FILTER_BUTTON_CLICK:
      const newActiveButton = {
        ...state.activeButton,
        [action.value]: !state.activeButton[action.value],
      };

      const newElements = state.elements.map((item) => {
        if (item.group === action.value) {
          return { ...item, active: !item.active };
        }
        return { ...item };
      });

      return {
        ...state,
        elements: [...newElements],
        activeButton: { ...newActiveButton },
      };
    case SHOW_SPINNER:
      return {
        ...state,
        loading: true,
      };
    case HIDE_SPINNER:
      return {
        ...state,
        loading: false,
      };
    case SEARCH_ELEMENT:
      const searchedElements = state.elements.map((item) => {
        if (action.value === '') {
          return { ...item, active: false };
        }
        if (item.name.includes(action.value)) {
          return { ...item, active: true };
        }
        if (!item.name.includes(action.value)) {
          return { ...item, active: false };
        }
        return { ...item };
      });
      return {
        ...state,
        elements: [...searchedElements],
      };
    case GET_PAGE_ID:
      return {
        ...state,
        activeElement: {
          ...state.activeElement,
          pageId: action.value,
        },
      };

    case GET_PAGE_CONTENT:
      return {
        ...state,
        activeElement: {
          ...state.activeElement,
          content: action.value,
        },
      };

    case CLEAR_DESCRIPTION:
      return {
        ...state,
        activeElement: {
          ...state.activeElement,
          pageId: 0,
          content: '',
        },
      };

    default:
      return state;
  }
}
