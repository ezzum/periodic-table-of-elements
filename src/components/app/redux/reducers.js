import { ELEMENTS_LOAD, ELEMENT_CLICK, ELEMENT_MOUSE_ENTER, FILTER_BUTTON_CLICK } from "./types";
import { initialState } from './initialState';

export function rootReducer(state = initialState, action) {
    switch (action.type) {
        case ELEMENTS_LOAD : 
            return {
                ...state, elements: action.value
            };
        case ELEMENT_MOUSE_ENTER : 
            return {
                ...state, previewElement: action.value
            };
        case ELEMENT_CLICK :
            return {
                ...state, activeElement: action.value
            };
        case FILTER_BUTTON_CLICK :
            
            const newActiveButton = {
                ...state.activeButton,
                [action.value] : !state.activeButton[action.value]
            }

            const newElements = state.elements.map(item => {
                if(item.group === action.value) {
                    return {...item, active: !item.active}
                }
                return {...item}
            })
            
            return {
                ...state,
                elements : [...newElements],
                activeButton : {...newActiveButton}

            }  
        default: return state;
    }
}