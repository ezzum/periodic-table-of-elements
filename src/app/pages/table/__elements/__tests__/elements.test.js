import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import ElementsGrid from "../elements-grid";
import React from "react";
import { ELEMENT_CLICK, ELEMENT_MOUSE_ENTER } from "../../../../redux/types";

jest.mock("../../__preview/preview", () => "div");
jest.mock("../../__filter-button/filter-button", () => "div");
jest.mock("react-router-dom", () => ({
  Link: "a",
  Route: ({ children, path }) => children({ match: path === "/description" }),
}));

describe("Elements tests", () => {
  const props = {
    ElementMouseEnter: jest.fn(),
    ElementClick: jest.fn(),
  };

  const initialStore = {
    elements: [
      {
        symbol: "H",
        number: 1,
        name: "водород",
        atomicMass: 1.007,
        column: 1,
        row: 1,
        group: "nonMetal",
      },
    ],
  };
  const mockStore = configureStore();
  let store, wrapper;

  beforeEach(() => {
    store = mockStore(initialStore);
    wrapper = mount(
      <Provider store={store}>
        <ElementsGrid {...props} />
      </Provider>
    );
  });

  it("snapshot test", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("click element test", () => {
    wrapper.find("a").simulate("click");

    expect(wrapper.props().store.getActions()).toEqual([
      { type: ELEMENT_CLICK, value: "водород" },
    ]);
  });

  it("mouse enter element test", () => {
    wrapper.find("a").simulate("mouseEnter");

    expect(wrapper.props().store.getActions()).toEqual([
      {
        type: ELEMENT_MOUSE_ENTER,
        value: {
          symbol: "H",
          number: 1,
          name: "водород",
          atomicMass: 1.007,
          column: 1,
          row: 1,
          group: "nonMetal",
        },
      },
    ]);
  });
});
