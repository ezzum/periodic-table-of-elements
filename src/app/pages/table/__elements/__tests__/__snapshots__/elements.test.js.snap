// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements tests snapshot test 1`] = `ReactWrapper {}`;
